### PR TITLE
SecretService as a proper class

### PR DESCRIPTION
**Pull request by JasonLG1979**
This makes SecretService a proper class and seems to also solve the issue mentioned in https://github.com/pithos/pithos/pull/390 by caching the password.

### CODE REVIEW

File: pithos/util.py [hunk 1]
```diff
@@ -24,50 +24,94 @@
     Gtk
 )
 
-_ACCOUNT_SCHEMA = Secret.Schema.new('io.github.Pithos.Account', Secret.SchemaFlags.NONE,
-                                    {"email": Secret.SchemaAttributeType.STRING})
+class _SecretService:
```
Comment (TingPing) [inline pithos/util.py]: __[flake8]__
* [E302] expected 2 blank lines, found 1

File: pithos/util.py [hunk 2]
```diff
@@ -24,50 +24,94 @@
...
-        if not num_items or default_collection not in unlocked:
-            _current_collection = Secret.COLLECTION_SESSION
-            logging.debug('The default keyring is locked. Using session collection.')
-        else:
-            logging.debug('The default keyring was unlocked.')
+        self._cached_password = password
+
+        return True
 
+SecretService = _SecretService()
```
Comment (TingPing) [inline pithos/util.py]: __[flake8]__
* [E305] expected 2 blank lines after class or function definition, found 1

File: pithos/util.py [hunk 3]
```diff
@@ -24,49 +24,95 @@
...
-        if not num_items or default_collection not in unlocked:
-            _current_collection = Secret.COLLECTION_SESSION
-            logging.debug('The default keyring is locked. Using session collection.')
-        else:
-            logging.debug('The default keyring was unlocked.')
+        self._cached_password = password
+
+        return True
+
+SecretService = _SecretService()
```
Comment (TingPing) [inline pithos/util.py]: __[flake8]__
* [E305] expected 2 blank lines after class or function definition, found 1

**Comment by TingPing** [comment]
I'm not sure this makes things much cleaner.

**Comment by JasonLG1979** [comment]
It encapsulates everything. It's better that having a bunch of related free functions just floating there isn't it? and as I said caching the password seems to solve the problem in 390. I think we are running into a situation where we're trying to get the password to fast after we set it. Caching it means we don't have to ask the secret service for it as much.

**Comment by JasonLG1979** [comment]
From a dogmatic python standpoint aside from caching the password we're trading globals for effectively a singleton. Both are viewed as bad but in this case a singleton is actually what we want.

File: pithos/util.py [hunk 4]
```diff
@@ -24,76 +24,101 @@
...
+            {'email': email},
+            self._current_collection,
+            'Pandora Account',
+            password,
+            None,
+        )
 
-    if not password:
-        return _clear_account_password(email)
+        if password_store_successful:
```
Comment (TingPing) [inline pithos/util.py]: __[flake8]__
* [W291] trailing whitespace

File: pithos/util.py [hunk 5]
```diff
@@ -24,76 +24,101 @@
...
+            None,
+        )
 
-    if not password:
-        return _clear_account_password(email)
+        if password_store_successful: 
+            self._cached_password = password
+        else:
+            self._cached_password = ''
+            logging.warning('Failed to store password')
```
Comment (TingPing) [inline pithos/util.py]: __[flake8]__
* [W291] trailing whitespace

**Comment by JasonLG1979** [comment]
@TingPing, This also seems to solve the problem where on 1st if I login to my Pandora One account Pithos quits. For the same reason as I mentioned above I'm guessing. I think we're asking for the password to fast after we set it.

**Comment by JasonLG1979** [comment]
I'd really like to see this and the ratings extension merged. I'm still playing with the Lasf.fm PR. We don't use much of pylast and the Last.fm api is pretty simple. I was toying with the idea of just ditching pylast and doing my own thing with Last.fm.

**Comment by TingPing** [comment]
We can't "cache" the password, its an external service that can change beyond our control and without notifying us.

**Comment by JasonLG1979** [comment]
> its an external service that can change beyond our control and without notifying us.

I'm not sure why anything would change the password set for and by Pithos besides Pithos?

**Comment by JasonLG1979** [comment]
The "external service" is apparently to slow. The alternative is to go async. But that's really tricky.

**Comment by JasonLG1979** [comment]
If we go async it gets worse. You set the password async and then ask for it async. There's a good chance they won't return in the right order.

**Comment by TingPing** [comment]
I believe the order of DBus messages is retained. We should being going async anyway.

**Comment by JasonLG1979** [comment]
OK I already have a branch where I started it.

**Pull request closed without merging**